File: endpoints/src/main/java/com/cromoteca/samples/nullable/endpoints/NullableEndpoint.java
package com.cromoteca.samples.nullable.endpoints;

import com.vaadin.flow.server.auth.AnonymousAllowed;
import dev.hilla.Endpoint;
import dev.hilla.Nonnull;
import java.util.List;
import java.util.Map;

@Endpoint
@AnonymousAllowed
public class NullableEndpoint {
  public Integer box(int n) {
    return n;
  }

  @Nonnull
  public String notNull() {
    return "";
  }

  public void nulls(Nulls argument) {}

  public void nonNullListOfNonNullStrings(@Nonnull List<@Nonnull String> argument) {}

  public void nullableArrayOfNullableStrings(String[] argument) {}

  public void nullableArrayOfNonNullStrings(@Nonnull String[] argument) {}

  public void nonNullArrayOfNullableStrings(String @Nonnull [] argument) {}

  public void nonNullArrayOfNonNullStrings(@Nonnull String @Nonnull [] argument) {}

  public static class Nulls {
    private @Nonnull Map<@Nonnull String, @Nonnull List<@Nonnull String>> allNull;
    private @Nonnull Map<@Nonnull String, @Nonnull List<String>> someNull;
    private @Nonnull List<@Nonnull Object> nullOrNot;

    public Map<String, List<String>> getAllNull() {
      return allNull;
    }

    public void setAllNull(@Nonnull Map<String, List<String>> allNull) {
      this.allNull = allNull;
    }

    public Map<String, List<String>> getSomeNull() {
      return someNull;
    }

    public void setSomeNull(Map<String, List<String>> someNull) {
      this.someNull = someNull;
    }

    // The type argument is nullable and cannot inherit the annotation since the type argument of
    // the field is different.
    public @Nonnull List<String> getNullOrNot() {
      return nullOrNot.stream().map(Object::toString).toList();
    }
  }
}

File: generator/src/test/resources/com/cromoteca/samples/nullable/endpoints/NullableEndpoint.ts
import client from '../../../../../connect-client.default.js';
import type Nulls from './NullableEndpoint/Nulls.js';
import { type EndpointRequestInit } from '@vaadin/hilla-frontend';

async function box(n: number, init?: EndpointRequestInit): Promise<number | undefined> {
    return client.call('com.cromoteca.samples.nullable.endpoints.NullableEndpoint', 'box', { n }, init);
}

async function nonNullArrayOfNonNullStrings(argument: Array<string>, init?: EndpointRequestInit): Promise<void> {
    return client.call('com.cromoteca.samples.nullable.endpoints.NullableEndpoint', 'nonNullArrayOfNonNullStrings', { argument }, init);
}

async function nonNullArrayOfNullableStrings(argument: Array<string | undefined>, init?: EndpointRequestInit): Promise<void> {
    return client.call('com.cromoteca.samples.nullable.endpoints.NullableEndpoint', 'nonNullArrayOfNullableStrings', { argument }, init);
}

async function nonNullListOfNonNullStrings(argument: Array<string>, init?: EndpointRequestInit): Promise<void> {
    return client.call('com.cromoteca.samples.nullable.endpoints.NullableEndpoint', 'nonNullListOfNonNullStrings', { argument }, init);
}

async function notNull(init?: EndpointRequestInit): Promise<string> {
    return client.call('com.cromoteca.samples.nullable.endpoints.NullableEndpoint', 'notNull', {}, init);
}

async function nullableArrayOfNonNullStrings(argument: Array<string> | undefined, init?: EndpointRequestInit): Promise<void> {
    return client.call('com.cromoteca.samples.nullable.endpoints.NullableEndpoint', 'nullableArrayOfNonNullStrings', { argument }, init);
}

async function nullableArrayOfNullableStrings(argument: Array<string | undefined> | undefined, init?: EndpointRequestInit): Promise<void> {
    return client.call('com.cromoteca.samples.nullable.endpoints.NullableEndpoint', 'nullableArrayOfNullableStrings', { argument }, init);
}

async function nulls(argument: Nulls | undefined, init?: EndpointRequestInit): Promise<void> {
    return client.call('com.cromoteca.samples.nullable.endpoints.NullableEndpoint', 'nulls', { argument }, init);
}

const NullableEndpoint = {
    box,
    nonNullArrayOfNonNullStrings,
    nonNullArrayOfNullableStrings,
    nonNullListOfNonNullStrings,
    notNull,
    nullableArrayOfNonNullStrings,
    nullableArrayOfNullableStrings,
    nulls,
};

export default NullableEndpoint;
